examples/migraphx/custom_op_hip_kernel/custom_op_hip_kernel.cpp

```diff
@@ -0,0 +1,91 @@
+#include <algorithm>
+#include <hip/hip_runtime.h>
+#include <migraphx/migraphx.hpp> // MIGraphX's C++ API
+#include <numeric>
+
+#define TENSOR_SIZE 3 * 3
+#define MIGRAPHX_HIP_ASSERT(x) (assert((x) == hipSuccess))
+/*
+ * Square each element in the array A and write to array C.
+ */
+template <typename T>
+__global__ void vector_square(T* C_d, const T* A_d, size_t N)
+{
+    size_t offset = (hipBlockIdx_x * hipBlockDim_x + hipThreadIdx_x);
+    size_t stride = hipBlockDim_x * hipGridDim_x;
+
+    for(size_t i = offset; i < N; i += stride)
+    {
+        C_d[i] = A_d[i] * A_d[i];
+    }
+}
+
+struct square_custom_op final : migraphx::experimental_custom_op_base
+{
+    virtual std::string name() const override { return "square_custom_op"; }
+    virtual migraphx::argument
+    compute(migraphx::context ctx, migraphx::shape, migraphx::arguments inputs) const override
+    {
+        // if compile options has offload_copy = true then, parameters and outputs will be
+        // automatically copied to and from GPUs' memory. Here assume that `inputs` arguments are
+        // already in the GPU, so no need to do Malloc, Free or Memcpy last element in the `inputs`
+        // is output argument, so it should be returned from compute method.
+        auto* input_buffer  = reinterpret_cast<float*>(inputs[0].data());
+        auto* output_buffer = reinterpret_cast<float*>(inputs[1].data());
+        MIGRAPHX_HIP_ASSERT(hipSetDevice(0));
+        const unsigned blocks          = 512;
+        const unsigned threadsPerBlock = 256;
+        hipLaunchKernelGGL(vector_square,
+                           dim3(blocks),
+                           dim3(threadsPerBlock),
+                           0,
+                           ctx.get_queue<hipStream_t>(),
+                           output_buffer,
+                           input_buffer,
+                           TENSOR_SIZE);
+        return inputs[1];
+    }
+    virtual migraphx::shape compute_shape(migraphx::shapes inputs) const override
+    {
+        assert(inputs.size() == 2);
+        assert(bool{inputs[0] == inputs[1]});
+        return inputs.back();
+    }
+};
+
+int main(int argc, const char* argv[])
+{
+    square_custom_op square_op;
+    migraphx::register_experimental_custom_op(square_op);
+    migraphx::program p;
+    migraphx::shape s{migraphx_shape_float_type, {3, 3}};
+    migraphx::module m = p.get_main_module();
+    auto x             = m.add_parameter("x", s);
+    auto neg_ins       = m.add_instruction(migraphx::operation("neg"), x);
+    // do allocation for the output buffer for the custom_kernel
+    auto alloc = m.add_instruction(
+        migraphx::operation("allocate",
+                            R"({"shape":{"type":"float_type","lens":[3, 3], "strides":[3, 1]}})"),
```

Comment:
We actually need a `m.add_allocation` method in our API to take a shape. There is no guarantee this will always be the same.

Reply:
Added `add_allocation`.